File: src/main/java/com/kamikazejam/datastore/command/sub/CmdCache.java
package com.kamikazejam.datastore.command.sub;

import com.kamikazejam.datastore.base.Cache;
import com.kamikazejam.datastore.base.Store;
import com.kamikazejam.datastore.command.SubCommand;
import org.bukkit.Bukkit;
import org.bukkit.command.Command;
import org.bukkit.command.CommandSender;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Arrays;
import java.util.List;

public class CmdCache extends SubCommand {
    public CmdCache() {}

    @Override
    public @NotNull List<String> getAliases() { return List.of("cache"); }
    @Override
    public @NotNull String getName() { return "cache"; }
    @Override
    public @Nullable String getPermission() { return "datastore.command.cache"; }
    @Override
    public @NotNull String getArgsDescription() { return "<cache>"; }

    @Override
    public boolean onCommand(CommandSender sender, Command cmd, String label, String[] args) {
        if (!sender.hasPermission(this.getPermission())) {
            this.sendNoPerm(sender);
            return true;
        }

        // TODO
        Bukkit.getLogger().info("Provided Sub Args for Cache: " + Arrays.toString(args));

        return true;
    }

//    @Override
//    public void perform() throws KamiCommonException {
//        Cache<?,?> cache = readArg();
//        List<String> localCacheKeys = getSomeKeyStrings(cache);
//
//        sender.sendMessage(Color.t("&7***** &6Store Cache: " + cache.getName() + " &7*****"));
//        sender.sendMessage(Color.t("&7" + cache.getLocalCacheSize() + " objects in local cache, first 10: " + Arrays.toString(localCacheKeys.toArray())));
//        sender.sendMessage(Color.t("&7Current State: " + (cache.isRunning() ? "&aRunning" : "&cNot running")));
//    }

    @SuppressWarnings("unchecked")
    private <K, X extends Store<X, K>> List<String> getSomeKeyStrings(Cache<?, ?> c) {
        Cache<K, X> cache = (Cache<K, X>) c;
        return cache.getLocalStore().getKeyStrings(cache).stream()
                .limit(10)
                .toList();
    }

    @Override
    public List<String> onTabComplete(CommandSender sender, Command cmd, String label, String[] args) {
        return List.of();
    }
}

File: src/main/kotlin/com/kamikazejam/datastore/command/sub/CmdCache.kt
package com.kamikazejam.datastore.command.sub

import com.kamikazejam.datastore.DataStoreAPI
import com.kamikazejam.datastore.base.Cache
import com.kamikazejam.datastore.base.Store
import com.kamikazejam.datastore.command.SubCommand
import com.kamikazejam.datastore.util.Color
import org.bukkit.command.CommandSender
import java.util.*

class CmdCache : SubCommand() {
    override val name: String
        get() = "cache"
    override val permission: String
        get() = "datastore.command.cache"
    override val argsDescription: String
        get() = "<cache>"

    override fun execute(sender: CommandSender, args: Array<String>) {
        if (args.isEmpty()) {
            this.sendUsage(sender)
            return
        }

        val cacheName = args[0]
        val cache = DataStoreAPI.getCache(cacheName)
        if (cache == null) {
            sender.sendMessage(Color.t("&cNo Cache matching \"$cacheName\". See `/datastore caches` for a list of them."))
            return
        }

        // Perform Command now that we have args
        val localCacheKeys = getSomeKeyStrings(cache)
        sender.sendMessage(Color.t("&7***** &6Store Cache: " + cache.name + " &7*****"))
        sender.sendMessage(
            Color.t(
                "&7" + cache.localCacheSize + " objects in local cache, first 10: " + localCacheKeys.toTypedArray()
                    .contentToString()
            )
        )
        sender.sendMessage(Color.t("&7Current State: " + (if (cache.running) "&aRunning" else "&cNot running")))
    }

    override fun getTabCompletions(sender: CommandSender, args: Array<String>): List<String> {
        val s = if (args.isEmpty()) "" else args[args.size - 1]
        return DataStoreAPI.caches.keys
            .stream()
            .filter { id: String -> id.lowercase(Locale.getDefault()).startsWith(s.lowercase(Locale.getDefault())) }
            .limit(20)
            .toList()
    }

    private fun <K, T : Store<T, K>> getSomeKeyStrings(c: Cache<K, T>): List<String> {
        return c.localStore.getKeyStrings(c).stream()
            .limit(10)
            .toList()
    }
}
